cirq-core/cirq/testing/routing_devices.py

```diff
@@ -0,0 +1,67 @@
+# Copyright 2021 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Provides test devices that can validate circuits during a routing procedure."""
+
+from typing import Hashable, Optional, Dict, TYPE_CHECKING
+
+import networkx as nx
+
+from cirq import devices, ops
+
+if TYPE_CHECKING:
+    import cirq
+
+
+class RoutingTestingDevice(devices.Device):
+    """Testing device to be used only for testing qubit connectivity in routing procedures."""
+
+    def __init__(self, nx_graph: nx.Graph, qubit_type: str = 'NamedQubit') -> None:
+        relabeling_map: Dict[Hashable, 'cirq.Qid'] = {}
+        if qubit_type == 'GridQubit':
+            relabeling_map = {old: devices.GridQubit(*old) for old in nx_graph}
+        elif qubit_type == 'LineQubit':
+            relabeling_map = {old: devices.LineQubit(old) for old in nx_graph}
+        else:
+            relabeling_map = {old: ops.NamedQubit(str(old)) for old in nx_graph}
+
+        # Relabel nodes in-place.
+        nx.relabel_nodes(nx_graph, relabeling_map, copy=False)
+
+        self._metadata = devices.DeviceMetadata(relabeling_map.values(), nx_graph)
+
+    @property
+    def metadata(self) -> Optional[devices.DeviceMetadata]:
+        return self._metadata
+
+    def validate_operation(self, operation: 'cirq.Operation') -> None:
+        for q in operation.qubits:
+            if q not in self._metadata.qubit_set:
+                raise ValueError(f'Qubit not on device: {q!r}.')
+
+        if len(operation.qubits) == 2 and operation.qubits not in self._metadata.nx_graph.edges:
```

Comment:
This will fail for an undirected graph with edges (1, 2) if you check whether qubits (2, 1) are on the device or not? Also, you need to raise if there's an operation on > 2 qubits (except maybe measurement operation, which is allowed to be on any subset of qubits).

Reply:
> This will fail for an undirected graph with edges (1, 2) if you check whether qubits (2, 1) are on the device or not?

I've just added some tests for this. It doesn't seem to be failing.

Reply:
> Also, you need to raise if there's an operation on > 2 qubits (except maybe measurement operation, which is allowed to be on any subset of qubits).

Working on it!